File: java/Trees/Diameter_of_Binary_Tree.java
package java.Trees;

class TreeNode {
  int val;
  TreeNode left, right;
}

public class Diameter_of_Binary_Tree {
  int diameter = 0;

  public int diameterOfBinaryTree(TreeNode root) {
    depth(root);
    return diameter;
  }

  int depth(TreeNode node) {
    if (node == null)
      return 0;
    int left = depth(node.left);
    int right = depth(node.right);
    diameter = Math.max(diameter, left + right);
    return 1 + Math.max(left, right);
  }
}

File: C++/Tree/Diameter_of_Binary_Tree.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
  int diameter = 0;

  int depth(TreeNode *root)
  {
    if (!root)
      return 0;

    int left = depth(root->left);
    int right = depth(root->right);

    diameter = max(diameter, left + right);

    return 1 + max(left, right);
  }

public:
  int diametrOfBinaryTree(TreeNode *root)
  {
    depth(root);
    return diameter;
  }
};